File: Day6_Majority Element II/majority_element.py
# https://www.geeksforgeeks.org/batch/gfg-160-problems/track/arrays-gfg-160/problem/majority-vote
def findMajority(arr):
    # Your Code goes here.
    count_1, count_2 = 0,0
    candidate_1, candidate_2 = 0,0
    for num in arr: # morries voting algorithm
        if count_1 == 0 and num != candidate_2:
            candidate_1,count_1 = num, 1
        elif count_2 == 0 and num != candidate_1:
            candidate_2,count_2 = num, 1
        elif num == candidate_1:
            count_1 += 1
        elif num == candidate_2:
            count_2 += 1
        else:
            count_1 -= 1
            count_2 -= 1

    result = []
    threshold = len(arr) // 3

    count_1 = count_2 = 0
    for num in arr:
        if candidate_1 == num:
            count_1 += 1
        elif candidate_2 == num:
            count_2 += 1

    if count_1 > threshold and count_2 > threshold:
        return [min(candidate_1, candidate_2), max(candidate_1, candidate_2)]
    elif count_1 > threshold:
        return [candidate_1]
    elif count_2 > threshold:
        return [candidate_2]
    else:
        return []

user_input = input("Enter space-separated integers: ")
arr = list(map(int, user_input.split()))
print(findMajority(arr=arr))

File: Day6_Majority Element II/majority_element.java
import java.util.ArrayList;
import java.util.Scanner;

public class majority_element {
public static ArrayList<Integer> findMajority(ArrayList<Integer> arr) {
    int candidate_1 = 0, candidate_2 = 0, count_1 = 0, count_2 = 0;
    for (int num : arr) {
        if(count_1 == 0 && num != candidate_2) {
            count_1 = 1;
            candidate_1 = num;
        } 
        else if(count_2 == 0 && num != candidate_1) {
            count_2 = 1;
            candidate_2 = num;
        }
        else if(num == candidate_1) 
            count_1++;
        else if(num == candidate_2)
            count_2++;
        else
            count_1--; count_2--;
    }

    count_1 = 0; count_2 = 0;
    int threshold = arr.size()/3;
    for (int num : arr) {
        if(num == candidate_1) 
            count_1++;
        else if(num == candidate_2)
            count_2++;
    }
    ArrayList<Integer> result = new ArrayList<>();
    if(count_1 > threshold && count_2 > threshold) {
        result.add(Math.min(candidate_1, candidate_2));
        result.add(Math.max(candidate_1,candidate_2));
        return result;
    }
    if(count_1 > threshold)
        result.add(candidate_1);
    else if(count_2 > threshold)
        result.add(candidate_2);
    return result;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.println("Enter the size and element of array: ");
        int n = sc.nextInt();
        ArrayList<Integer> arr = new ArrayList<Integer>();
        for (int i = 0; i < n; i++) {
            arr.add(sc.nextInt());
        }
        ArrayList<Integer> result = findMajority(arr);
        for (int num : result) {
            System.out.print(num + " ");
        }
        System.out.println();
        sc.close();
    }
}
